repair/repair.cc

```diff
@@ -1025,7 +1029,19 @@ static future<> repair_range(repair_info& ri, const dht::token_range& range) {
         return do_for_each(ri.cfs.begin(), ri.cfs.end(), [&ri, &neighbors, range] (auto&& cf) {
             ri._sub_ranges_nr++;
             if (ri.row_level_repair()) {
-                return repair_cf_range_row_level(ri, cf, range, neighbors);
+                if (ri.dropped_tables.count(cf)) {
+                    return make_ready_future<>();
```

Comment:
The whole thing is vulerable to drop/create races. If we drop a table and re-create it with the same name, will repair notice? Note the new table can have a completely different schema.

We should work with table uuids, not names. A drop/create will result in a new uuid even if they have the same name. It's also better for rename, if we ever support it.

Reply:
No. Repair loops over the ranges it asked to repair, for each range, it loops over the tables. So, if the a table is dropped and a new table is created with the same name, it is possible for some ranges the repair is performed with the old tables and some with the new tables. 

However, user specifies the table name not the uuid when user starts the repair. It is not wrong to repair the new tables for some of the ranges.  What can we do in this case? I think it is worse to fix the whole ranges with the old tables when the old table is dropped. 

Btw, for a given range, repair master and followers will always repair the same table because they repair the same schema id.

Reply:
Do we have a dtest of what happens if during a repair the table is dropped and/or recreated during a repair?
Because I'm not seeing the big picture here (I guess the big picture isn't in this patch) - does a repair use the table name, or its UUID, to communicate with its inner loop, and also with repair slaves? It would be good to have a dtest to demonstrate that. If the situation is already correct in this respect (is it?) it would be shame for this patch to break it.

Reply:
We do have a dtest to test the table is dropped during repair: update_cluster_layout_tests.py:TestUpdateClusterLayout.simple_add_new_node_while_schema_changes_test. 

Repair master and followers already use the same schema version for a given table and range we repair. 

The problem here is what do we do if a table is dropped during repair.  Should we fail the whole bootstrap operation because the table is dropped and the repair failed (bootstrap uses repair now)?

Reply:
The new table can have an incompatible schema (different partition key and clustering key types). It can only end in disaster.

To avoid the race, convert the table names to uuids exactly once. If the drop/create happens before the lookup, we repair the new table. If it happens during or after the lookup, we won't repair (but that's fine, since it's a new table that did not exist when repair started).

Reply:
If a table is dropped during repair, we should ignore it. A dropped table is fully synchronized.

Reply:
Yes, new table can have incompatible schema. But currently, it won't cause any problem because repair master and followers use the same schema id for a given range we repair. 

I do agree that it is more consistent to ask all ranges inside a repair request to use the same schema id too. I will add this change.

Reply:
Ok, so we ignore dropped tables and report the repair as success.